jfix-stdlib-ratelimiter/src/test/kotlin/ru/fix/stdlib/ratelimiter/SuspendableRateLimitedDispatcherTest.kt

```diff
@@ -0,0 +1,713 @@
+        }
+        reporter.buildReportAndReset().assertSoftly {
+            it.getMetric("active_async_operations") shouldBe 2
+        }
+
+        trackableDispatcher.submitCompletedTasks(7..10)
+        await().atMost(Duration.ofSeconds(1)).until {
+            trackableDispatcher.isSubmittedTaskInvoked(7..10)
+        }
+        reporter.buildReportAndReset().assertSoftly {
+            it.getMetric("active_async_operations") shouldBe 2
+        }
+
+        await().atMost(Duration.ofSeconds(3)).until {
+            trackableDispatcher.isSubmittedTaskFinished(1..10)
+        }
+        reporter.buildReportAndReset().assertSoftly {
+            it.getMetric("active_async_operations") shouldBe 0
+        }
+
+        trackableDispatcher.completeAllAndClose()
+    }
+
+    /**
+     * When dispatcher closingTimeout is enough for pending tasks to complete
+     * such tasks will complete normally
+     */
+    @Test
+    @Timeout(10, unit = TimeUnit.SECONDS)
+    fun `on shutdown fast tasks with CompletableFuture complete normally`() {
+
+        val dispatch = createDispatcher(closingTimeout = 5_000)
+
+        val futures = List(3) {
+            dispatch.compose {
+                completedFuture(true)
+            }
+        }
+        dispatch.close()
+
+        futures.forEach { future: CompletableFuture<*> ->
+            future.isDone.shouldBeTrue()
+            future.isCompletedExceptionally.shouldBeFalse()
+        }
+    }
+
+    /**
+     * When dispatcher closingTimeout is enough for pending tasks to complete
+     * such tasks will complete normally
+     */
+    @Test
+    @Timeout(10, unit = TimeUnit.SECONDS)
+    fun `on shutdown fast tasks complete normally`() = runBlocking {
+
+        val dispatch = createDispatcher(closingTimeout = 5_000)
+
+        val results = List(3) {
+            dispatch.decorateSuspend {
+                true
+            }
+        }
+        dispatch.close()
+
+        results.forEach { result ->
+            result.shouldBeTrue()
+        }
+    }
+
+    @Test
+    @Timeout(5, unit = TimeUnit.SECONDS)
+    fun `on shutdown slow tasks with CompletableFuture complete exceptionally`() {
+        val dispatch = createDispatcher(closingTimeout = 0)
+
+        for (i in 1..3) {
+            scope.async {
+                dispatch.compose {
+                    sleep(3000)
+                    completedFuture(true)
+                }
+            }
+        }
+
+        val exc = shouldThrow<RejectedExecutionException> {
+            dispatch.close()
+        }
+        exc.message!!.contains("timeout while waiting for coroutines finishing").shouldBeTrue()
+    }
+
+    @Test
+    @Timeout(5, unit = TimeUnit.SECONDS)
+    fun `on shutdown slow tasks complete exceptionally`() = runBlocking {
+        val dispatch = createDispatcher(closingTimeout = 0)
+
+        for (i in 1..3) {
+            scope.async {
+                dispatch.decorateSuspend {
+                    sleep(3000)
```

Comment:
do not use blocking calls

```suggestion
                    delay(3000)
```

Reply:
Done